### ISSUE TITLE
Error when linking ww3_grib

### ISSUE THREAD

**Issue by Josue9740** [NOAA-EMC/WW3#542]
### **Describe the bug**
I am trying to compile a program with a modified version of the switch `NCEP_glwu`, in which grib files are used. In the modified version of `NCEP_glwu` I simply deleted the switches `NOPA` and `LRB4`, since their use raises an error like this: 

`sort_switch: could not process all switches [NOPA LRB4] abort` 

I do not know how to solve this error either, although this is not the main topic of this issue. 

Using the modified `NCEP_glwu switch`, in which `NOPA` and `LRB4` have been removed, I always get an error message during the compilation procedure when `ad3`, after processing `ww3_grib`, tries to link this file (`ww3_grib`).  This is the error: 

```
Processing ww3_grib
---------------------
ad3 : processing ww3_grib
         Linking ww3_grib
make: *** [makefile:400: /home/tesalia/WW3/model/exe/ww3_grib] Error 1
```

I checked the makefile located at `WW3/model/src`. Line 400 in `makefile` states:

`@$(aPb)/link ww3_grib w3parall wmmdatmd w3triamd  w3gdatmd w3wdatmd w3adatmd w3idatmd w3odatmd     w3sdb1md w3sbt1md  w3src4md w3flx1md w3snl3md    w3iogrmd w3iogomd constants w3servmd w3timemd w3arrymd w3dispmd w3gsrumd`

I also checked the ad3 file, located at `WW3/model/bin` and I saw that `Error 1` refers to `Input errors`. Therefore, I checked the `ww3_grib` file that should be located at `WW3/model/exe` and it does not exist. In other words, it is not created by my model. I think this may be the source of the problem, but I do not know how to handle it. Below I describe my configuration. 

Error files (especially `ww3_grib.err`) and logs located at `WW3/model/tmp` are empty. The only file with some content is located at `WW3/model/obj_SEQ/link.err`, and states:
```
gfortran: error: big_endian: No existe el archivo o el directorio
gfortran: error: byterecl: No existe el archivo o el directorio
gfortran: error: unrecognized command line option ‘-convert’; did you mean ‘-fconvert=’?
gfortran: error: unrecognized command line option ‘-assume’
gfortran: error: unrecognized command line option ‘-prec-div’
gfortran: error: unrecognized command line option ‘-prec-sqrt’
gfortran: error: unrecognized command line option ‘-ip’; did you mean ‘-p’?
```

I do not know how to handle this information.

### **To Reproduce**
Since I have other programs using some libraries required for WW3 (e.g., I have install a WRF model that also uses jasper, zlib and png), I am not using a hpc-stack nor something similar. I have installed almost all libraries by hand. 

- Next, I detail the libraries that were installed in my system before installing WW3:

**NetCDF**
```
nf-config --version -> Output: netCDF-Fortran 4.5.2
nf-config --has-nc4-> Output: yes
```
**Jasper** -> Jasper 1.900.1
**HDF5** -> HDF5 1.10.4
**zlib** -> 1.2.11
**libpng** -> 1.6.37
**libjpeg** -> 8.2.2

- I installed these other libraries (code is described too): 

**CMake** (Please, see [https://github.com/Kitware/CMake/tree/21e60da5bb0715e09f2feba5930ea2c8bc8180af)](https://github.com/Kitware/CMake/tree/21e60da5bb0715e09f2feba5930ea2c8bc8180af) -> Version 3.22.2021
```
git clone https://github.com/Kitware/CMake.git
cd CMake
./bootstrap --prefix=/home/tesalia/cmake -- -DCMAKE_USE_OPENSSL=OFF
./bootstrap && make && sudo make install 
```
**NCEPLIBS** (it includes several libraries. Please, see [https://github.com/NOAA-EMC/NCEPLIBS)](https://github.com/NOAA-EMC/NCEPLIBS)
```
git clone https://github.com/NOAA-EMC/NCEPLIBS
mkdir -p build && cd build
cmake –DCMAKE_INSTALL_PREFIX=/usr/lib/x86_64-linux-gnu/nceplibs /home/tesalia/NCEPLIBS
sudo make
```
**WAVEWATCH 3** (I am using my own fork, but the library is exactly the same, saving from some modified switches as indicated above)
```
sudo git clone https://github.com/Josue9740/WW3
cd WW3
sudo sh model/bin/ww3_from_ftp.sh
```

- I declared these environment variables in ~/.bash_profile
```
export PNG_LIB=/usr/lib/x86_64-linux-gnu/libpng.a
export Z_LIB=/usr/lib/x86_64-linux-gnu/libz.a
export JASPER_LIB=/usr/lib/x86_64-linux-gnu/libjasper.a
export G2_LIB4=/usr/local/g2-3.4.1/lib/libg2_4.a
export W3NCO_LIB4=/usr/local/w3nco-2.4.1/lib/libw3nco_4.a 
export BACIO_LIB4=/usr/local/bacio-2.4.1/lib/libbacio_4.a
```

- Model compilation
```
sudo ./model/bin/w3_setup model -c gnu -s NCEP_glwu
cd model/bin
sudo ./w3_automake
```

As indicated above, NCEP_glwu is a modified version without the switches `NOPA` and `LRB4`. I am using gnu compiler since if I use the same code but replacing `-c gnu` by `-c ntel` I get the next error:
```
/home/tesalia/WW3/model/bin/ad3: línea 215: ifort: command not found
      *** w3adc error ***
          file constants.F90 not found (2)
```

In any case, if I use the `gnu` option I also get the error indicated above.

**Comment by JessicaMeixner-NOAA** [comment]
The NOPA and LRB4 switches have been deprecated.  

For ww3_grib linking issues you usually just need to make sure the variables here: https://github.com/NOAA-EMC/WW3/blob/develop/model/bin/link.tmpl#L111 are defined and that the libraries you are stating in this line exist. 

I'm not sure about ww3_grib and w3_automake (maybe @mickaelaccensi knows), but if you switch DIST and MPI to SHRD (and remove any OMP) you should just be able to use w3_make and that has works for me.

**Comment by Josue9740** [comment]
@JessicaMeixner-NOAA, first of all, thank you for your quick response.

Switching `DIST` and `MPI` to `SHRD`, removing any `OMP`, and using `make` instead of `automake` has contributed to improve the compilation of the model, but only to some extent. Before these changes, I got the output indicated above:
```
Processing ww3_grib
---------------------
ad3 : processing ww3_grib
         Linking ww3_grib
make: *** [makefile:400: /home/tesalia/WW3/model/exe/ww3_grib] Error 1
```
Right now, the output is slightly different. I get the same error (in a different line of makefile as I am using a different switch configuration), but the program seems to be able to process a part:

```
Processing ww3_grib
---------------------
ad3 : processing constants
ad3 : processing w3servmd
ad3 : processing w3gsrumd
ad3 : processing w3arrymd
ad3 : processing w3gdatmd
ad3 : processing w3idatmd
ad3 : processing w3odatmd
ad3 : processing w3adatmd
ad3 : processing w3dispmd
ad3 : processing w3timemd
ad3 : processing wmmdatmd
ad3 : processing w3sdb1md
ad3 : processing w3sbt4md
ad3 : processing w3flx1md
ad3 : processing w3src4md
ad3 : processing w3triamd
ad3 : processing w3parall
ad3 : processing w3snl1md
ad3 : processing w3wdatmd
ad3 : processing w3iogomd
ad3 : processing w3iogrmd
ad3 : processing ww3_grib
      Linking ww3_grib
make: *** [makefile:368: /home/tesalia/WW3/model/exe/ww3_grib] Error 1
```

In addition, the message in the `link.err` file located at `WW3/model/obj_SEQ` is the same again:
```
gfortran: error: big_endian: No existe el archivo o el directorio
gfortran: error: byterecl: No existe el archivo o el directorio
gfortran: error: unrecognized command line option ‘-convert’; did you mean ‘-fconvert=’?
gfortran: error: unrecognized command line option ‘-assume’
gfortran: error: unrecognized command line option ‘-prec-div’
gfortran: error: unrecognized command line option ‘-prec-sqrt’
gfortran: error: unrecognized command line option ‘-ip’; did you mean ‘-p’?
```

I have checked the libraries you indicated in the code and, as you can see in the attached picture, the corresponding environment variables are correctly defined. Is there any example in the `regtests` directory that uses GRIB? I have checked most of the examples but did not find anyone. Perhaps, could you provide me with the example switch file you were using? 

Thank you so much for your help!

![Captura](https://user-images.githubusercontent.com/86227863/142607227-bdb09fdb-3b6d-4321-b4bb-d15929a6d186.PNG)

**Comment by JessicaMeixner-NOAA** [comment]
When you do ./w3_setup are you choosing -c gnu?   

Also, since all the libraries exist, go ahead an do a:
 ls $PNG_LIB
to make sure that library does in fact exist. 

A regression test that uses grib is for example: 
./bin/run_test  -s MPI -w work_c -m grdset_c -f -p <FILLIN> -n 24  -o all ../model ww3_ufs1.1
where you need to replace <FILLIN> with srun, mpirun, or whatever mpi command you use on your machine.

**Comment by Josue9740** [comment]
Hi again, Jessica.

Yes, I am choosing `-c gnu`. I have checked what you say regarding the libraries. I found some problems in the initial environment variables' declaration. I fixed the problem, and I checked again that the variables have been properly defined by using `ls $LIBNAME` (e.g., `ls $Z_LIB`). Now, the variable declaration seems to be correct, but the same error is still there.

I have read some information on the Internet that may be of help for this case. The initial error that I found in the `link.err` file was:
```
gfortran: error: big_endian: File or directory not found
gfortran: error: byterecl: File or directory not found
gfortran: error: unrecognized command line option ‘-convert’; did you mean ‘-fconvert=’?
gfortran: error: unrecognized command line option ‘-assume’
gfortran: error: unrecognized command line option ‘-prec-div’
gfortran: error: unrecognized command line option ‘-prec-sqrt’
gfortran: error: unrecognized command line option ‘-ip’; did you mean ‘-p’?
```
Then, based on the information I found I changed line 111 in the file link.tmpl from:
`opt="$opt -convert big_endian -assume byterecl -prec-div -prec-sqrt -ip"`
To:
`opt="$opt -fconvert=big-endian -assume byterecl -prec-div -prec-sqrt -ip"`

As a result, the error message has changed from the one indicated above to the next one (the first line has dissappear):
```
gfortran: error: byterecl: File or directory not found
gfortran: error: unrecognized command line option ‘-convert’; did you mean ‘-fconvert=’?
gfortran: error: unrecognized command line option ‘-assume’
gfortran: error: unrecognized command line option ‘-prec-div’
gfortran: error: unrecognized command line option ‘-prec-sqrt’
gfortran: error: unrecognized command line option ‘-ip’; did you mean ‘-p’?
```
I tried to change the remaining parts of line 111 (i.e., `-assume byterecl`, `-prec-div`, `-prec-sqrt`, `-ip`), but I do not find the proper synthax to do so. Does this change mean something to you?

Best regards

**Comment by Josue9740** [comment]
> 

Dear  @JessicaMeixner-NOAA, 

I have also run the example you provide me with the next code:

`sudo ./bin/run_test -s MPI -w ../model_examples/work_ww3_ufs1.1 -m grdset_c -f -p mpirun -n 24 -o all ../model ww3_ufs1.1`

The next error arises:
```
+--------------------+
|  Grid preprocessor |
+--------------------+


                *****************************
              ***   WAVEWATCH III make      ***
                *****************************

Main directory    : /home/tesalia/WW3/model
Scratch directory : /home/tesalia/WW3/model/tmp
Save source codes : yes
Save listings     : yes
Shared Object     : No

[INFO] list of the programs selected : ww3_grid

Processing ww3_grid
---------------------
ad3 : processing w3gdatmd
ad3 : processing scrip_kindsmod [no w3adc]
mv: cannot stat 'scrip_kindsmod.o': No such file or directory
mv: cannot stat 'w3gdatmd.o': No such file or directory
ad3 : processing scrip_iounitsmod [no w3adc]
ad3 : processing scrip_timers [no w3adc]
ad3 : processing scrip_constants [no w3adc]
ad3 : processing w3odatmd
mv: cannot stat 'scrip_iounitsmod.o': No such file or directory
mv: cannot stat 'scrip_constants.o': No such file or directory
mv: cannot stat 'scrip_timers.o': No such file or directory
mv: cannot stat 'w3odatmd.o': No such file or directory
ad3 : processing w3idatmd
ad3 : processing scrip_errormod [no w3adc]
ad3 : processing w3timemd
ad3 : processing w3dispmd
mv: cannot stat 'scrip_errormod.o': No such file or directory
mv: cannot stat 'w3idatmd.o': No such file or directory
mv: cannot stat 'w3timemd.o': No such file or directory
ad3 : processing w3nmlgridmd
mv: cannot stat 'w3dispmd.o': No such file or directory
ad3 : processing scrip_netcdfmod [no w3adc]
ad3 : processing wmmdatmd
ad3 : processing w3flx1md
mv: cannot stat 'scrip_netcdfmod.o': No such file or directory
mv: cannot stat 'w3nmlgridmd.o': No such file or directory
ad3 : processing w3adatmd
ad3 : processing w3src4md
mv: cannot stat 'w3flx1md.o': No such file or directory
mv: cannot stat 'wmmdatmd.o': No such file or directory
ad3 : processing scrip_grids [no w3adc]
mv: cannot stat 'w3src4md.o': No such file or directory
mv: cannot stat 'w3adatmd.o': No such file or directory
mv: cannot stat 'scrip_grids.o': No such file or directory
ad3 : processing w3triamd
ad3 : processing w3snl1md
ad3 : processing w3parall
ad3 : processing scrip_remap_vars [no w3adc]
mv: cannot stat 'w3snl1md.o': No such file or directory
mv: cannot stat 'w3parall.o': No such file or directory
mv: cannot stat 'w3triamd.o': No such file or directory
mv: cannot stat 'scrip_remap_vars.o': No such file or directory
ad3 : processing w3iogrmd
ad3 : processing w3wdatmd
ad3 : processing scrip_remap_write [no w3adc]
ad3 : processing scrip_remap_conservative [no w3adc]
mv: cannot stat 'w3iogrmd.o': No such file or directory
mv: cannot stat 'w3wdatmd.o': No such file or directory
mv: cannot stat 'scrip_remap_conservative.o': No such file or directory
mv: cannot stat 'scrip_remap_write.o': No such file or directory
ad3 : processing scrip_remap_read [no w3adc]
mv: cannot stat 'scrip_remap_read.o': No such file or directory
ad3 : processing scrip_interface
mv: cannot stat 'scrip_interface.o': No such file or directory
ad3 : processing wmscrpmd
mv: cannot stat 'wmscrpmd.o': No such file or directory
ad3 : processing w3gridmd
mv: cannot stat 'w3gridmd.o': No such file or directory
ad3 : processing ww3_grid
mv: cannot stat 'ww3_grid.o': No such file or directory
: no valid option
make: *** [makefile:43: /home/tesalia/WW3/model/exe/ww3_grid] Error 1

ERROR: Error occured during WW3 ww3_grid build
```

There is an error message in some files contained into the path `WW3/model/tmp`, which states: 
`/home/tesalia/WW3/model/bin/comp: line 98: -c: command not found`

**Comment by aliabdolali** [comment]
you need to add `-c gnu` in the command line

`sudo ./bin/run_test -s MPI -c gnu -w ../model_examples/work_ww3_ufs1.1 -m grdset_c -f -p mpirun -n 24 -o all ../model ww3_ufs1.1`

**Comment by Josue9740** [comment]
Dear @aliabdolali, thank you so much for your reply.

I am afraid that the problem is still there even if I add -c gnu to my code.

**Comment by aliabdolali** [comment]
this is what the log file shows:
```
There is an error message in some files contained into the path WW3/model/tmp, which states:
/home/tesalia/WW3/model/bin/comp: line 98: -c: command not found
```

**Comment by Josue9740** [comment]
Thank you for your quick response, @aliabdolali.

This is my first time working with WAVEWATCH III, so let me know if I am wrong, but I think that the error message  is not referred to the command line `sudo ./bin/run_test -s MPI -w ../model_examples/work_ww3_ufs1.1 -m grdset_c -f -p mpirun -n 24 -o all ../model ww3_ufs1.1`. 

- First, if it was the case, I would get the error inmediately after introducing the command line stated above, and not some time later when the model is to be compiled. 
- Second, the model is supposed to use a compiler by default (gnu) if you do not specify one, isn't it?
- Third, I think that the error message `-c: command not found` refers to line 98 in file `comp` (`/home/tesalia/WW3/model/bin/comp`).  Line 98 in this file states: `$comp $opt $name.$fext 1> $name.out 2> $name.err`, and some lines above (line 71) it is declared `opt="-c -J$path_m -g -fno-second-underscore -ffree-line-length-none -fconvert=big-endian -O3 -march=native"`. 

I think this later line is the source of the error, but I do not know why it occurs or how to solve it. I have tried by changing the output (for instance, using only `grib` or only `netCDF`) and I always get the same error. However, if I use the native output the model works perfectly. Could this fact be related with some issue regarding the libraries?

**Comment by Josue9740** [comment]
Dear all,

None of the previous solutions worked. I have declared the environment variables and I did check they exist, among many other checks. In fact, I have added additional environment variables I read about on different forums. I have discovered that my system is also running the WRF model, which was installed some time ago. There seems to be some kind of incompatibility between the libraries. Finally, I handle this problem by constructing the WW3 program into a Docker. It is working perfectly right now. If someone needs help to construct the Docker can contact me.

Thank everyone for your help.
